### PR TITLE
docs, dev-requirements: upgrade sphinx and try to fix some type issues with documentation

### PR DESCRIPTION
**Pull request by Exirel**
### Description

First, I upgraded Sphinx to the latest version, which requires to upgrade Furo as well. That went well so that's good! Then I added links to the general index and the Python module index, because they are nice.

Then I tried to fix Sphinx autodoc's warnings regarding type annotations. I more or less figure out a way to fix one or two... but then I got stuck very hard: Sphinx, currently, does not handle type annotation properly. For some reason it works for `TypedRule` now but not with some `datetime.datetime` annotation? There are clearly bugs (that are not new to version 7.0 of Sphinx by the way), and I tried few options (such as using `autodoc_type_aliases`, or for a moment I switch to another autodoc extension, with the same issues). Sopel is not an isolated case, [as this issue shows it](https://github.com/sphinx-doc/sphinx/issues/9813).

So yeah, since it's mostly Sphinx's fault, I decided to give up on the remaining warnings. I don't think we can do much: we could, hypothetically, tinker for hours with the code... but I don't see the value here. I already spent quite a few hours on this and I'm not happy so let's not waste any more time.

**Note**: this won't work with Python 3.7 as I use a feature from Python 3.8 (`typing.Protocol`). So until @dgw publish his PR to remove Python 3.7 support, this will be a draft only.

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

File: sopel/tools/target.py [hunk 1]
```diff
@@ -1,16 +1,15 @@
+"""Trigger's targets: users and channels."""
```
Comment (dgw) [inline sopel/tools/target.py:1]: Not much to do with triggers. These objects are mostly used for keeping track of which users and channels (plus privileges) Sopel knows about (in `bot.channels` and `bot.users`).

```suggestion
"""User and channel objects used in state tracking."""
```

File: dev-requirements.txt [hunk 2]
```diff
@@ -6,11 +6,11 @@ flake8-future-import
 flake8-import-order
 flake8-type-checking; python_version >= '3.8'
 # Sphinx theme
-furo==2022.4.7
+furo==2023.5.20
 pytest~=7.1.0
 pytest-vcr~=1.0.2
 requests-mock~=1.9.3
-sphinx>=4,<5
+sphinx
```
Comment (dgw) [inline dev-requirements.txt:13]: Should specify at least `>=7.0` to agree with the minimum Sphinx version from `docs/source/conf.py`.

**Review by dgw — CHANGES_REQUESTED** [review]
Line notes are self-explanatory. Besides a CodeQL warning about `SopelIdentifierMemory` not overriding `__eq__()` (not actually new, but has been triggered nonetheless by touching that class's code), I thought it would be useful to point out the Sphinx warnings that are still present in the current version of this PR, courtesy of [Netlify's last build log](https://app.netlify.com/sites/sopel-unstable-docs/deploys/64d7734968482b0008bf660d#L529-L545).

<details>
<summary>Click for Sphinx warnings</summary>

```
4:56:47 AM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage:1: WARNING: py:class reference target not found: ModeDetails
4:56:47 AM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage:1: WARNING: py:class reference target not found: PrivilegeDetails
4:56:47 AM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage:1: WARNING: py:class reference target not found: ModeTuple
4:56:47 AM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage.ignored_modes:1: WARNING: py:class reference target not found: ModeTuple
4:56:47 AM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage.modes:1: WARNING: py:class reference target not found: ModeDetails
4:56:47 AM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage.privileges:1: WARNING: py:class reference target not found: PrivilegeDetails
4:56:47 AM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.parse_modestring:1: WARNING: py:class reference target not found: ModeTuple
4:56:47 AM: /opt/build/repo/sopel/plugins/handlers.py:docstring of sopel.plugins.handlers.AbstractPluginHandler.get_capability_requests:1: WARNING: py:class reference target not found: plugin_decorators.capability
4:56:47 AM: /opt/build/repo/sopel/plugins/handlers.py:docstring of sopel.plugins.handlers.PyModulePlugin.get_capability_requests:1: WARNING: py:class reference target not found: plugin_decorators.capability
4:56:47 AM: /opt/build/repo/sopel/plugins/rules.py:docstring of sopel.plugins.rules.AbstractRule.is_channel_rate_limited:1: WARNING: py:class reference target not found: datetime
4:56:47 AM: /opt/build/repo/sopel/plugins/rules.py:docstring of sopel.plugins.rules.AbstractRule.is_user_rate_limited:1: WARNING: py:class reference target not found: datetime
4:56:47 AM: /opt/build/repo/sopel/plugins/rules.py:docstring of sopel.plugins.rules.Rule.is_channel_rate_limited:1: WARNING: py:class reference target not found: datetime
4:56:47 AM: /opt/build/repo/sopel/plugins/rules.py:docstring of sopel.plugins.rules.Rule.is_user_rate_limited:1: WARNING: py:class reference target not found: datetime
4:56:47 AM: /opt/build/repo/sopel/tools/identifiers.py:docstring of sopel.tools.identifiers.Identifier.casemapping:1: WARNING: py:class reference target not found: sopel.tools.identifiers.Casemapping
4:56:47 AM: /opt/build/repo/sopel/tools/memories.py:docstring of sopel.tools.memories.SopelIdentifierMemory.make_identifier:1: WARNING: py:class reference target not found: sopel.tools.identifiers.IdentifierFactory
4:56:47 AM: /opt/build/repo/sopel/tools/target.py:docstring of sopel.tools.target.Channel.make_identifier:1: WARNING: py:class reference target not found: IdentifierFactory
4:56:47 AM: /opt/build/repo/sopel/trigger.py:docstring of sopel.trigger.Trigger:1: WARNING: py:class reference target not found: Match
```

</details>

_Since I'm unsure how long Netlify keeps build logs around, or whether the deploy preview logs are public, I also pasted the relevant portion here._

File: dev-requirements.txt [hunk 3]
```diff
@@ -6,11 +6,11 @@ flake8-future-import
 flake8-import-order
 flake8-type-checking; python_version >= '3.8'
 # Sphinx theme
-furo==2022.4.7
+furo==2023.5.20
```
[the remark this answers is not in the document]
Comment (Exirel) [reply dev-requirements.txt:9]: IIRC furo has sphinx requirements? I don't remember exactly, but I think my pip complained about something. But you're right, I'll upgrade that!

[the remark this answers is not in the document]
Comment (dgw) [reply dev-requirements.txt:9]: > IIRC furo has sphinx requirements?

`>=6.0,<8.0` at present: https://github.com/pradyunsg/furo/blob/7aeb6d204f808ae0b0da8a5dee6f6035d9cd04cf/pyproject.toml#L20

**Comment by dgw** [comment]
Was looking for something else (after someone asked a question in Furo's discussions) and noticed https://deploy-preview-2496--sopel-unstable-docs.netlify.app/package/db#sopel.db.SopelDB has a not-very-helpful annotation of a parameter:

image

(It's the `<class 'sopel.tools.identifiers.Identifier'>` notation. It's too verbose, and it's not even x-ref'd.)

Along the way to that, I also noticed some long signatures and remembered about sphinx-doc/sphinx#11011 adding a configuration setting (`maximum_signature_line_length`) in Sphinx 7.1.0 that makes long signatures multi-line, just like we do in the code (and [Furo should handle it fine](https://github.com/pradyunsg/furo/discussions/388#discussioncomment-5665098)). Want to make that part of this patch, since you're upgrading Sphinx anyway?

File: docs/source/conf.py [hunk 4]
```diff
@@ -24,7 +21,7 @@
 # -- General configuration -----------------------------------------------------
 
 # If your documentation needs a minimal Sphinx version, state it here.
-needs_sphinx = '4.0'
+needs_sphinx = '7.0'
```
[the remark this answers is not in the document]
Comment (dgw) [reply docs/source/conf.py:24]: (My latest top-level comment/idea about possibly enabling multi-line for signatures that are too long will require bumping this to 7.1.)

**Comment by dgw** [comment]
Uh-oh. We'd better keep an eye on this. "Furo is broken on Sphinx 7.2" https://github.com/pradyunsg/furo/discussions/693

**Comment by dgw** [comment]
Quick responses! https://github.com/pradyunsg/furo/discussions/693#discussioncomment-6753443 (Furo fix) + https://github.com/pradyunsg/furo/discussions/693#discussioncomment-6753629 (Sphinx fix)

**Comment by Exirel** [comment]
> Along the way to that, I also noticed some long signatures and remembered about [sphinx-doc/sphinx#11011](https://github.com/sphinx-doc/sphinx/pull/11011) adding a configuration setting (`maximum_signature_line_length`) in Sphinx 7.1.0 that makes long signatures multi-line, just like we do in the code (and [Furo should handle it fine](https://github.com/pradyunsg/furo/discussions/388#discussioncomment-5665098)). Want to make that part of this patch, since you're upgrading Sphinx anyway?

Hell yeah, and it works!

**Comment by dgw** [comment]
Hmm, I wonder why solving the Sphinx requirement failed… There's a [new Furo version](https://pradyunsg.me/furo/changelog/#zesty-zaffre) (2023.9.10) you can use as an excuse to re-run CI 😁

**Comment by SnoopJ** [comment]
> Hmm, I wonder why solving the Sphinx requirement failed…

I believe it's because the relevant requirement is `sphinx<8,>=7.2.3`, but Sphinx (surprisingly?) [dropped 3.8 support in 7.2.0](https://www.sphinx-doc.org/en/master/changes.html#release-7-2-0-released-aug-17-2023)

**Comment by dgw** [comment]
> surprisingly?

A bit. 3.8 isn't EOL until over a year from now (October '24), _and_ it's also weird (to me) to drop a Python version in a minor bump.

That said, SemVer is meaningless… but oy. I guess we can probably get away with Sphinx 7.2.x for Python >=3.9 and Sphinx 7.1.x for Python 3.8. We Probably™ don't rely on any of the fixes/changes in Sphinx 7.2.0, and Furo should have finally sorted out all of its cross-version compatibility at this point.

I hope. 😅

**Comment by dgw** [comment]
@Exirel I looked at the changes since my last review and I _think_ this is fine except for the failure on Python 3.8. Could you sort that out when you get a moment, so we can verify that the docs build correctly? Looks like Netlify didn't even try to make a deploy preview for the latest commit; there's no check annotation. ☹️

File: dev-requirements.txt [hunk 5]
```diff
@@ -6,11 +6,12 @@ flake8-future-import
...
 flake8-type-checking; python_version >= '3.8'
 # Sphinx theme
-furo==2022.4.7
+furo==2023.9.10
 pytest~=7.1.0
 pytest-vcr~=1.0.2
 requests-mock~=1.9.3
-sphinx>=4,<5
+sphinx>=7.1.0,<8; python_version <= '3.8'
+sphinx>=7.2.0<8; python_version > '3.8'
```
Comment (dgw) [inline dev-requirements.txt:14]: How did CI pass with this? 😂

```suggestion
sphinx>=7.2.0,<8; python_version > '3.8'
```

File: sopel/plugin.py [hunk 6]
```diff
@@ -121,30 +121,15 @@ class CapabilityNegotiation(enum.Enum):
...
-class capability:
-    """Decorate a function to request a capability and handle the result.
+class CapabilityHandler(Protocol):
+    """:class:`~typing.Protocol` definition for capability handler.
 
-    :param name: name of the capability to negotiate with the server; this
-                 positional argument can be used multiple times to form a
-                 single ``CAP REQ``
-    :param handler: optional keyword argument, acknowledgement handler
+    When a plugin requests a capability using the :class:`capability`, it can
```
Comment (dgw) [inline sopel/plugin.py:127]: ``:class:`capability` type``?

**Review by dgw — COMMENTED** [review]
I think this qualifies as nitpicking now. Just force-push it; as long as you don't also rebase, I'll be able to use the "force-pushed" event to view what changed for re-review purposes. 😸

**Comment by dgw** [comment]
Before I re-review this one last time and give you the :shipit: squash instruction, are the [Sphinx warnings in the build log](https://app.netlify.com/sites/sopel-unstable-docs/deploys/651340d7e8108000087a84e6#L521-L537) for cdc9a10bbfd68c58d01028bcbdf4575d42b01837 part of what you were trying to fix?

<details>
<summary>Long Sphinx log snippet</summary>

```
3:37:31 PM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage:1: WARNING: py:class reference target not found: ModeDetails
3:37:31 PM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage:1: WARNING: py:class reference target not found: PrivilegeDetails
3:37:31 PM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage:1: WARNING: py:class reference target not found: ModeTuple
3:37:31 PM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage.ignored_modes:1: WARNING: py:class reference target not found: ModeTuple
3:37:31 PM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage.modes:1: WARNING: py:class reference target not found: ModeDetails
3:37:31 PM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.ModeMessage.privileges:1: WARNING: py:class reference target not found: PrivilegeDetails
3:37:31 PM: /opt/build/repo/sopel/irc/modes.py:docstring of sopel.irc.modes.parse_modestring:1: WARNING: py:class reference target not found: ModeTuple
3:37:31 PM: /opt/build/repo/sopel/plugins/handlers.py:docstring of sopel.plugins.handlers.AbstractPluginHandler.get_capability_requests:1: WARNING: py:class reference target not found: plugin_decorators.capability
3:37:31 PM: /opt/build/repo/sopel/plugins/handlers.py:docstring of sopel.plugins.handlers.PyModulePlugin.get_capability_requests:1: WARNING: py:class reference target not found: plugin_decorators.capability
3:37:31 PM: /opt/build/repo/sopel/plugins/rules.py:docstring of sopel.plugins.rules.AbstractRule.is_channel_rate_limited:1: WARNING: py:class reference target not found: datetime
3:37:31 PM: /opt/build/repo/sopel/plugins/rules.py:docstring of sopel.plugins.rules.AbstractRule.is_user_rate_limited:1: WARNING: py:class reference target not found: datetime
3:37:31 PM: /opt/build/repo/sopel/plugins/rules.py:docstring of sopel.plugins.rules.Rule.is_channel_rate_limited:1: WARNING: py:class reference target not found: datetime
3:37:31 PM: /opt/build/repo/sopel/plugins/rules.py:docstring of sopel.plugins.rules.Rule.is_user_rate_limited:1: WARNING: py:class reference target not found: datetime
3:37:31 PM: /opt/build/repo/sopel/tools/identifiers.py:docstring of sopel.tools.identifiers.Identifier.casemapping:1: WARNING: py:class reference target not found: sopel.tools.identifiers.Casemapping
3:37:31 PM: /opt/build/repo/sopel/tools/memories.py:docstring of sopel.tools.memories.SopelIdentifierMemory.make_identifier:1: WARNING: py:class reference target not found: sopel.tools.identifiers.IdentifierFactory
3:37:31 PM: /opt/build/repo/sopel/tools/target.py:docstring of sopel.tools.target.Channel.make_identifier:1: WARNING: py:class reference target not found: IdentifierFactory
3:37:31 PM: /opt/build/repo/sopel/trigger.py:docstring of sopel.trigger.Trigger:1: WARNING: py:class reference target not found: Match
```

</details>

I ask because at least some of the reference targets not found are part of your diff, as aliases in `docs/source/conf.py`.

**Comment by Exirel** [comment]
Yeah, I tried, but as I said in the PR's description, for some reasons Sphinx just can't handle some cases, and I still don't understand why or how. I gave up on these. 😭

**Comment by Exirel** [comment]
All squashed and ready to go.

**Review by dgw — APPROVED** [review]
I've just looked over all the changes again, finally. (Made a good break from esoteric stuff like #2525, not gonna lie.) This patch has been in-flight more than long enough already! :shipit:

While there were a couple of little wording changes I would otherwise ask to make, it's on me for missing them through all the previous reviews. Little documentation tweaks are great… to ignore until it's time to start working on 8.0.1.

**Pull request merged**